Add tests for StandardScaler covering with_mean and with_std

agoscinski opened the pull request:
I hope it is clear from the commit messages and title. I can also split it up into two PRs, if requested.

- add tests for StandardScaler covering with_mean and with_std
- adds gradient_compliant flag to StandardScaler

Luthaf commented on src/equisolve/numpy/preprocessing/_base.py:
```diff
@@ -39,6 +48,7 @@ def __init__(
         column_wise: bool = False,
         rtol: float = 0.0,
         atol: float = 1e-12,
+        gradient_compliant: bool = True,
```
Why do we need this flag? I don't see a use case for `gradient_compliant=False` at all, so it feels like this is only adding needless complexity =/

agoscinski replied:
It goes back to the discussion of https://github.com/lab-cosmo/equistore/issues/118
I am not sure, I would like to try it out. Its not much added complexity in terms of code and it just makes the user aware of a complexity that he should be anyway aware of. So in the worst case this is not used and it only makes the user aware of what is happening here.

Luthaf replied:
I would apply [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) here. If we don't have a use case for it, do not add it, even if it is a simple change, because a lot of unnecessary little changes can lead to death by a thousand cuts. 

If you want to play with it, this change does not have to be in the main branch, it can stay in a personal branch until we find a usecase for it.

> makes the user aware of what is happening here.

I would rather do this with documentation on the `with_mean` parameter, saying there that only the values are modified.

agoscinski replied:
Fair enough, I thought about it and I cannot think of any transformer benefiting of this option, and if it turns out to be useful somewhere else, we can *then* think about.

Luthaf commented on src/equisolve/numpy/preprocessing/_base.py:
```diff
@@ -121,9 +130,7 @@ def fit(
-                    mean_values = np.zeros(
-                        self.n_properties_.block(key).values.flatten()
-                    )
+                    mean_values = np.zeros(self.n_properties_)
```
unrelated to this PR, but why is the StandardScaler not implemented in terms of equistore operations? It looks like it is re-doing what should be `equistore.mean`, `equistore.std`, `equistore.substract`, `equistore.div`. Is something missing in the operations?

agoscinski replied:
because these were not implemented when we implemented the class

Luthaf replied:
fair enough 😄 I'll open an issue to update this =)

agoscinski replied:
made an issue #18

Luthaf commented on tests/numpy/preprocessing/test_base.py:
```diff
@@ -25,6 +25,9 @@ class TestStandardScaler:
     n_strucs = 5
+
+    # this computation is only needed because equistore does not offer random
+    # initialization TODO change as soons as equistore as random initialization
```
What do you mean with random initialization? something like `np.rand((3, 4))`? I doubt we will be able to provide something like this, since we also need metadata =/ You could have an helper function here that create a tensormap with known metadata and random data, in the same way this works: https://github.com/lab-cosmo/equistore/blob/master/python/tests/utils.py#L6

agoscinski replied:
Both things seems useful. For now I would just use an adapted version of the file you linked here, and put it in a test utils foder as you did in equstore.

agoscinski commented:
Changes addressing the review are in this commit
[addressing comments of PR review](https://github.com/lab-cosmo/equisolve/pull/15/commits/9b3542de24ec3547398e59267bbe111704238fa3)

After approval I rebase this into two commits, first one adding the tests/numpy/utils.py file and the second commit with the rest, but I think this way is easier for reviewing.

agoscinski commented on src/equisolve/numpy/preprocessing/_base.py:
```diff
@@ -177,26 +168,19 @@ def fit(
-            for parameter in self.parameter_keys:
-                if parameter == "values":
-                    continue
+            for parameter in X_block.gradients_list():
```
this change was required because we allow to only apply the standardizer on certain gradient keys using parameter_keys, so if parameter_keys does not contain all keys, later the returned TensorBlock only contained the gradients in paramater_keys.  That was not the intention, so we just the means to zero and scale to 1 when parameter is not in parameter_keys

Luthaf replied:
This makes me think the usage of `parameter_keys` is a bit fragile. I guess this can be addressed when re-writing this class with operations, so it is fine for now.

Could you remind me why this functionality is not built on top of `remove_gradients`? Is there something missing?

agoscinski replied:
We thought that a user might for example want to only fit on energies, but wants to also predict forces (for a quick and dirty model) without removing the gradients, because we still want to output the force error at the end. I think it can be discussed how useful this actually is, but I wouldn't touch it in this PR.

Luthaf commented on tests/numpy/preprocessing/test_base.py:
```diff
@@ -6,89 +6,99 @@
-ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
+# temporary solution importing test utils till we find a better solution
+# for relative imports, see issue #20
+sys.path.append("../")
+from utils import random_single_block_no_components_tensor_map  # noqa
```
Do you want to fix #20 before merging this? Should be relatively easy

Luthaf reviewed (approved):
This is good to go for me, if you want to clean it up a bit and fix #20 first go for it, otherwise it will happen in separate PR!

PicoCentauri commented on src/equisolve/numpy/preprocessing/_base.py:
```diff
@@ -205,7 +188,7 @@ def fit(
 
-                if self.with_std:
+                if self.with_std and (parameter in self.parameter_keys):
                     X_mean = np.average(X_mat, weights=sample_weights, axis=0)
                     var = np.average((X_mat - X_mean) ** 2, axis=0)
```
Maybe it is actually useful to use the `mean_over_samples` and `variance_over_samples` function here.

agoscinski replied:
adding to issue https://github.com/lab-cosmo/equisolve/issues/18
